File: user_service.py
import sqlite3
from typing import Optional, Tuple

from db import get_connection
from auth import hash_password, verify_password


def create_user(username: str, email: str, password: str) -> Tuple[bool, str]:
    """
    Create a new user account.
    Returns (success, message).
    """

    if not all([username, email, password]):
        return False, "All fields are required"

    if len(password) < 8:
        return False, "Password must be at least 8 characters long"

    try:
        with get_connection() as conn:
            cur = conn.cursor()
            cur.execute(
                """
                INSERT INTO users (username, email, password_hash)
                VALUES (?, ?, ?)
                """,
                (username.strip(), email.strip(), hash_password(password))
            )
        return True, "User created successfully"

    except sqlite3.IntegrityError:
        return False, "Username or email already exists"

    except Exception:
        return False, "Unable to create user at this time"


def authenticate_user(username: str, password: str) -> Optional[int]:
    """
    Authenticate a user.
    Returns user_id on success, None otherwise.
    """

    if not username or not password:
        return None

    try:
        with get_connection() as conn:
            cur = conn.cursor()
            cur.execute(
                """
                SELECT id, password_hash
                FROM users
                WHERE username = ?
                """,
                (username.strip(),)
            )
            row = cur.fetchone()

        if not row:
            # Prevent user enumeration timing attacks
            verify_password(password, hash_password("dummy-password"))
            return None

        if verify_password(password, row["password_hash"]):
            return row["id"]

        return None

    except Exception:
        return None

def get_user_email(user_id):
    with get_connection() as conn:  # use `with`!
        cur = conn.cursor()
        cur.execute("SELECT email FROM users WHERE id = ?", (user_id,))
        row = cur.fetchone()
        if row:
            return row["email"]
        return None


def delete_user(user_id: int) -> bool:
    """Delete user and all associated data. Returns True if deleted."""
    try:
        with get_connection() as conn:
            cur = conn.cursor()
            cur.execute("DELETE FROM users WHERE id = ?", (user_id,))
            return cur.rowcount > 0
    except Exception as e:
        print("Error deleting user:", e)
        return False
    
def user_exists(user_id):
    """Check if a user exists in the database."""
    with get_connection() as conn:
        cur = conn.cursor()
        cur.execute("SELECT 1 FROM users WHERE id = ?", (user_id,))
        return cur.fetchone() is not None

File: frontend/userService.js
const { db } = require("../backend/db");
const { hashPassword, verifyPassword } = require("../backend/auth");

function createUser(username, email, password) {
  if (!username || !email || !password)
    return { ok: false, msg: "All fields are required" };
  if (password.length < 8)
    return { ok: false, msg: "Password must be at least 8 characters long" };

  try {
    const stmt = db.prepare(
      "INSERT INTO users (username, email, password_hash) VALUES (?, ?, ?)"
    );
    const info = stmt.run(
      username.trim(),
      email.trim(),
      hashPassword(password)
    );
    if (info.changes === 1)
      return {
        ok: true,
        msg: "User created successfully",
        user_id: info.lastInsertRowid,
      };
    return { ok: false, msg: "Failed to create user" };
  } catch (e) {
    if (e && e.code === "SQLITE_CONSTRAINT_UNIQUE")
      return { ok: false, msg: "Username or email already exists" };
    console.error("Create user error:", e);
    return { ok: false, msg: "Unable to create user at this time" };
  }
}

function authenticateUser(username, password) {
  if (!username || !password) return null;
  try {
    const row = db
      .prepare("SELECT id, password_hash FROM users WHERE username = ?")
      .get(username.trim());
    if (!row) return null;
    if (verifyPassword(password, row.password_hash)) return row.id;
    return null;
  } catch (e) {
    console.error("Auth error:", e);
    return null;
  }
}

function deleteUser(user_id) {
  try {
    const info = db.prepare("DELETE FROM users WHERE id = ?").run(user_id);
    return info.changes > 0;
  } catch (e) {
    console.error("Delete user error:", e);
    return false;
  }
}

function userExists(user_id) {
  try {
    const row = db.prepare("SELECT id FROM users WHERE id = ?").get(user_id);
    return !!row;
  } catch (e) {
    return false;
  }
}

function getUserEmail(user_id) {
  try {
    const row = db.prepare("SELECT email FROM users WHERE id = ?").get(user_id);
    return row ? row.email : null;
  } catch (e) {
    console.error("getUserEmail error:", e);
    return null;
  }
}

module.exports = { createUser, authenticateUser, deleteUser, userExists };
